src/problems/bvp_problems.jl

```diff
@@ -78,50 +78,52 @@ every solve call.
 * `p`: The parameters for the problem. Defaults to `NullParameters`
 * `kwargs`: The keyword arguments passed onto the solves.
 """
-struct BVProblem{uType, tType, isinplace, P, F, bF, PT, K} <:
+struct BVProblem{uType, tType, isinplace, P, F, BF, PT, K} <:
        AbstractBVProblem{uType, tType, isinplace}
     f::F
-    bc::bF
+    bc::BF
     u0::uType
     tspan::tType
     p::P
     problem_type::PT
     kwargs::K
-    @add_kwonly function BVProblem{iip}(f::AbstractODEFunction, bc, u0, tspan,
+
+    @add_kwonly function BVProblem{iip}(f::AbstractBVPFunction{iip}, bc, u0, tspan,
         p = NullParameters(),
         problem_type = StandardBVProblem();
         kwargs...) where {iip}
         _tspan = promote_tspan(tspan)
         warn_paramtype(p)
         new{typeof(u0), typeof(_tspan), iip, typeof(p),
             typeof(f), typeof(bc),
-            typeof(problem_type), typeof(kwargs)}(f, bc, u0, _tspan, p,
+            typeof(problem_type), typeof(kwargs)}(f, f.bc, u0, _tspan, p,
             problem_type, kwargs)
     end
 
     function BVProblem{iip}(f, bc, u0, tspan, p = NullParameters(); kwargs...) where {iip}
-        BVProblem(ODEFunction{iip}(f), bc, u0, tspan, p; kwargs...)
+        BVProblem(BVPFunction{iip}(f, bc), bc, u0, tspan, p; kwargs...)
     end
 end
 
 TruncatedStacktraces.@truncate_stacktrace BVProblem 3 1 2
 
-function BVProblem(f::AbstractODEFunction, bc, u0, tspan, args...; kwargs...)
+function BVProblem(f::AbstractBVPFunction, bc, u0, tspan, args...; kwargs...)
     BVProblem{isinplace(f, 4)}(f, bc, u0, tspan, args...; kwargs...)
 end
 
 function BVProblem(f, bc, u0, tspan, p = NullParameters(); kwargs...)
-    BVProblem(ODEFunction(f), bc, u0, tspan, p; kwargs...)
+    BVProblem(BVPFunction(f, bc), bc, u0, tspan, p; kwargs...)
```

Comment:
If `BVPFunction` stores `bc` should be duplicate the storage in `BVProblem` again?

Reply:
Yeah, indeed, the problem construction process in ```BVProblem``` is similar with ```SDEProblem```, they both have duplicate ```bc``` in ```BVProblem``` and ```g``` in ```SDEProblem```, but if we want to have a ```BVPFunction``` stores both ```f``` and ```bc```, duplicating ```bc``` could really making the problem constructor concise. We need to note that when we are unpacking ```BVProblem```, we actually get a ```BVPFunction``` but not ```f```.

https://github.com/SciML/SciMLBase.jl/blob/4d926fccf201c68f8194c66e5580cec104bd047d/src/problems/sde_problems.jl#L130-L132

Reply:
In that case it's fair to stay consistent.

Maybe we can add an additional dispatch on BVPFunction where we automatically pull out the `bc` during problem construction that way user wont have to do `BVProblem(BVPFunction(f, bc), bc...)` and instead can specify `BVProblem(BVPFunction(f, bc)....)`

Reply:
I think you mean a dispatch on ```BVProblem``` right? Just updated and now we can directly specify ```BVProblem(BVPFunction(f, bc).....)``` to construct a ```BVProblem```.
Do we need to do the same for ```SDEProblem``` and ```SDEFunction```? The problem construct of ```SDEProblem``` also need users to do somthing like ```SDEProblem(SDEFunction(f, g), g)```, see here: https://docs.sciml.ai/DiffEqDocs/stable/tutorials/sde_example/#Using-Higher-Order-Methods

Reply:
@ChrisRackauckas  do you think this is a valid API choice? Is there a particular reason other Problem Types don't have this?

Reply:
That was done so it would be non-breaking. But I'd be happy to just change the SDEProblem one at this point since it has effectively deprecated the `prob.g` a long time back.

Reply:
Updated both ```BVProblem``` and ```SDEProblem```.

Reply:
It looks like the change of ```SDEProblem``` is a breaking one.

Reply:
Sure, I am working on fixing StochasticDiffEq.jl errors

Reply:
yes